File: packages/python/src/bd_offline_geocoder/place.py
"""Locality and postal-code helpers from OSM city tags + unique thana codes."""

from __future__ import annotations

import json
import re
from functools import lru_cache
from importlib import resources
_SKIP = frozenset(
    {
        "bangladesh",
        "dhaka",
        "dhaka city",
        "dhaka city corporation",
        "dhaka north city corporation",
        "dhaka south city corporation",
    }
)
_SPLIT = re.compile(r"[,/|]")


def normalize_place_name(value: str) -> str:
    cleaned = " ".join(value.strip().casefold().split())
    if "(" in cleaned:
        cleaned = cleaned.split("(", 1)[0].strip()
    return cleaned


def locality_from_city(city: str | None, *skip_names: str | None) -> str | None:
    if not isinstance(city, str) or not city.strip():
        return None
    skip = set(_SKIP)
    for name in skip_names:
        if isinstance(name, str) and name.strip():
            skip.add(normalize_place_name(name))
            skip.add(name.strip().casefold())
    for part in _SPLIT.split(city):
        cleaned = part.strip()
        if not cleaned:
            continue
        if cleaned.casefold() in skip or normalize_place_name(cleaned) in skip:
            continue
        return cleaned
    return None


def lookup_postcode(district: str | None, *names: str | None) -> str | None:
    if not district:
        return None
    table = _postcode_table().get(normalize_place_name(district))
    if not table:
        return None
    for name in names:
        if not isinstance(name, str) or not name.strip():
            continue
        folded = name.strip().casefold()
        if folded.startswith("ward") or "city corporation" in folded:
            continue
        key = normalize_place_name(name)
        if key in _SKIP:
            continue
        code = table.get(key)
        if code:
            return code
    return None


def postcode_candidates(
    locality: str | None,
    city: str | None,
    union: str | None,
    upazila: str | None,
) -> list[str]:
    names: list[str] = []
    for value in (locality, city, union, upazila):
        if isinstance(value, str) and value.strip():
            names.append(value.strip())
            for part in _SPLIT.split(value):
                cleaned = part.strip()
                if cleaned and cleaned not in names:
                    names.append(cleaned)
    return names


@lru_cache(maxsize=1)
def _postcode_table() -> dict[str, dict[str, str]]:
    path = resources.files("bd_offline_geocoder.data.postcodes").joinpath(
        "unique_area_postcodes.json"
    )
    with path.open("r", encoding="utf-8") as file:
        payload = json.load(file)
    districts = payload.get("districts")
    return districts if isinstance(districts, dict) else {}


_ROAD_PREFIX = re.compile(r"^(?:road|rd)\s*[#:]?\s*(.+)$", re.IGNORECASE)
_NAMED_ROAD_NUMBER = re.compile(
    r"^(.*?\broad)\s+([\d০-৯][\d০-৯/\-]*)$", re.IGNORECASE
)
_NUMERIC_ROAD = re.compile(r"^[\d০-৯][\d০-৯/\-]*$")


def is_city_corporation(name: str | None) -> bool:
    return isinstance(name, str) and "city corporation" in name.casefold()


def split_road(road: str | None) -> tuple[str | None, str | None]:
    """Return ``(road_number, road_name)`` from an OSM street string."""
    if not isinstance(road, str) or not road.strip():
        return None, None
    value = road.strip()
    prefixed = _ROAD_PREFIX.match(value)
    if prefixed:
        return prefixed.group(1).strip(), None
    if _NUMERIC_ROAD.fullmatch(value):
        return value, None
    named = _NAMED_ROAD_NUMBER.match(value)
    if named:
        return named.group(2).strip(), named.group(1).strip()
    return None, value


def thana_upazila_name(upazila: str | None, area_village: str | None) -> str | None:
    if isinstance(upazila, str) and upazila.strip() and not is_city_corporation(upazila):
        return upazila.strip()
    if not isinstance(area_village, str) or not area_village.strip():
        return None
    area = area_village.strip()
    folded = area.casefold()
    if "thana" in folded or "upazila" in folded:
        return area
    return f"{area} Thana"


def city_name(upazila: str | None, district: str | None) -> str | None:
    if is_city_corporation(upazila) and isinstance(district, str) and district.strip():
        return district.strip()
    return None

File: packages/flutter/lib/src/place.dart
import 'area_postcodes.dart';

const _skip = {
  'bangladesh',
  'dhaka',
  'dhaka city',
  'dhaka city corporation',
  'dhaka north city corporation',
  'dhaka south city corporation',
};

final _split = RegExp(r'[,/|]');

String normalizePlaceName(String value) {
  var cleaned = value.trim().toLowerCase().replaceAll(RegExp(r'\s+'), ' ');
  if (cleaned.contains('(')) {
    cleaned = cleaned.split('(').first.trim();
  }
  return cleaned;
}

String? localityFromCity(String? city, [List<String?> skipNames = const []]) {
  if (city == null || city.trim().isEmpty) return null;
  final skip = {..._skip};
  for (final name in skipNames) {
    if (name == null || name.trim().isEmpty) continue;
    skip
      ..add(normalizePlaceName(name))
      ..add(name.trim().toLowerCase());
  }
  for (final part in city.split(_split)) {
    final cleaned = part.trim();
    if (cleaned.isEmpty) continue;
    if (skip.contains(cleaned.toLowerCase()) ||
        skip.contains(normalizePlaceName(cleaned))) {
      continue;
    }
    return cleaned;
  }
  return null;
}

String? lookupPostcode(String? district, List<String?> names) {
  if (district == null || district.trim().isEmpty) return null;
  final table = areaPostcodes[normalizePlaceName(district)];
  if (table == null) return null;
  for (final name in names) {
    if (name == null || name.trim().isEmpty) continue;
    final folded = name.trim().toLowerCase();
    if (folded.startsWith('ward') || folded.contains('city corporation')) {
      continue;
    }
    final key = normalizePlaceName(name);
    if (_skip.contains(key)) continue;
    final code = table[key];
    if (code != null) return code;
  }
  return null;
}

List<String> postcodeCandidates({
  String? locality,
  String? city,
  String? union,
  String? upazila,
}) {
  final names = <String>[];
  for (final value in [locality, city, union, upazila]) {
    if (value == null || value.trim().isEmpty) continue;
    names.add(value.trim());
    for (final part in value.split(_split)) {
      final cleaned = part.trim();
      if (cleaned.isNotEmpty && !names.contains(cleaned)) {
        names.add(cleaned);
      }
    }
  }
  return names;
}

bool isCityCorporation(String? name) =>
    name != null && name.toLowerCase().contains('city corporation');

({String? roadNumber, String? roadName}) splitRoad(String? road) {
  if (road == null || road.trim().isEmpty) {
    return (roadNumber: null, roadName: null);
  }
  final value = road.trim();
  final prefixed = RegExp(r'^(?:road|rd)\s*[#:]?\s*(.+)$', caseSensitive: false)
      .firstMatch(value);
  if (prefixed != null) {
    return (roadNumber: prefixed.group(1)!.trim(), roadName: null);
  }
  if (RegExp(r'^[\d০-৯][\d০-৯/\-]*$').hasMatch(value)) {
    return (roadNumber: value, roadName: null);
  }
  final named = RegExp(r'^(.*?\broad)\s+([\d০-৯][\d০-৯/\-]*)$', caseSensitive: false)
      .firstMatch(value);
  if (named != null) {
    return (roadNumber: named.group(2)!.trim(), roadName: named.group(1)!.trim());
  }
  return (roadNumber: null, roadName: value);
}

String? thanaUpazilaName(String? upazila, String? areaVillage) {
  if (upazila != null && upazila.trim().isNotEmpty && !isCityCorporation(upazila)) {
    return upazila.trim();
  }
  if (areaVillage == null || areaVillage.trim().isEmpty) return null;
  final area = areaVillage.trim();
  final folded = area.toLowerCase();
  if (folded.contains('thana') || folded.contains('upazila')) return area;
  return '$area Thana';
}

String? cityName(String? upazila, String? district) {
  if (isCityCorporation(upazila) && district != null && district.trim().isNotEmpty) {
    return district.trim();
  }
  return null;
}
